### DISCUSSION TITLE
S3 bucket versioning changing filename

### DISCUSSION THREAD

**Discussion by HEAVYPOLY** [pocketbase/pocketbase#3545] [Q&A]
It seems like the filename is changed each time it is overwritten on pocketbase, so instead of creating a new version on connected S3 it creates a new file in the bucket.  Is this how S3 versioning is supposed to work?  What is a good way to do versioning with pb / S3?
<img width="1702" alt="image" src="https://github.com/pocketbase/pocketbase/assets/24432677/84cd8740-346b-497a-876a-e65775e6553e">

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.
PocketBase doesn't support file (or any other record data) versioning out of the box.

**Reply by HEAVYPOLY** [reply]
Ok, then would it be possible to have an option where uploading to S3 keeps the same filename?  Currently it seems the filename changes each time the file is overwritten. (the random string after canvas_ changes.  is that generated automatically by pb?)

this is the formdata I am submitting to pb:
```javascript
  const form = new FormData()
  const dateTime = getUTCDateTIme()
  if (!canvas.dateCreated) {
    canvas.dateCreated = dateTime
  }
  form.append('id', id)
  form.append('file', new Blob([canvas.file]), 'canvas.wbp')
  form.append('image', new Blob([canvas.image]), 'image.webp')
  form.append('owner', canvas.owner)
  form.append('lastSynced', canvas.lastSynced)
  form.append('dateCreated', canvas.dateCreated)
  await pb.collection('canvases').update(id, form)
  ```

**Reply by ganigeorgiev** [reply]
> Ok, then would it be possible to have an option where uploading to S3 keeps the same filename?

But why, what is the use case for this?

Each uploaded file is treated as a new entry. I'll try to simplify the file management internals with the planned refactoring in the near future but there are no plans for now to allow "replacing" files with the same name using the web apis.

**Comment by HEAVYPOLY** [comment]
To allow S3 file versioning like this under one filename:
<img width="1182" alt="image" src="https://github.com/pocketbase/pocketbase/assets/24432677/9bece854-765d-4646-aa5e-39cbc2626043">

https://youtu.be/eRIq-8GRfas?si=r2dGi4oWRXgtvVKQ&t=389

Although the current way with each file treated as a new entry also works.  I should probably disable versioning on S3 because I'm getting sort of double versioning with S3 trying to version each new entry.

**Reply by ganigeorgiev** [reply]
But it doesn't really make sense to have versioning on the S3 side because the PocketBase record will not be aware of the old file versions, aka. you'll never be able to access or restore the old files via the PocketBase APIs. Additionally, not all S3 providers have retention/version policies.

**Comment by HEAVYPOLY** [comment]
Ok, the current way makes more sense then and I'll disable S3 versioning.  Is there a way to get a list of all the pb generated filenames by date?

**Reply by ganigeorgiev** [reply]
I'm failing to understand your use case and what you are trying to do.

Please understand that I have literally no context about your use case and questions like "Can I do X" without any clarification are hard to provide a concrete answer for because it could range from "it depends" to "why do you want to do it". There could be also another more efficient approach to your problem.

In any case, there is no web API that will return list with all uploaded files from all collections.
You can emulate this, by creating a "media" collection and associate all files to it, and then use the regular collection list/view apis to retrieve the files.

If you want programmatically to list all S3 uploaded files, you can call [`filesystem.List(prefix)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.19.0/tools/filesystem#System.List), something like (_but note that this could be very slow if you have a lot of files_):

```go
fsys, err := app.NewFilesystem()
defer fsys.Close()

objects, err := fsys.List(somePrefix)
```

**Comment by HEAVYPOLY** [comment]
My usecase:  I am developing a painting app, the user has canvases that they are painting to, editing and saving over with autosave.  I want a versioning system so they can view and recover previous versions of their canvas.  Something similar to how google docs works or dropbox versioning.

Would pb multi file array be a good way to do this?  
looking for functionality similar to this:  https://docs.aws.amazon.com/AmazonS3/latest/userguide/RetrievingObjectVersions.html
https://docs.aws.amazon.com/code-library/latest/ug/s3_example_s3_ListObjectVersions_section.html
<img width="400" alt="webbytownLayerClipArrrivesUI" src="https://github.com/pocketbase/pocketbase/assets/24432677/2ba023b0-87bb-4815-8c89-7407d570ee2d">

**Comment by HEAVYPOLY** [comment]
Bump for this.

I need a way to access the files that have been uploaded (s3) to a record's single file field.

canvasRecord.file (single file field) may have many files uploaded to it, each overwriting the last.  Is there a way to access the list of previously uploaded files and their timestamps?  So that the user could roll back to one of the older versions if necessary

**Reply by ganigeorgiev** [reply]
As mentioned previously, we don't keep old files and there is no files versioning out of the box.
It is also not possible to modify the autogenerated file name and I don't understand why do you need it and how adding a timestamp will help you.

If I understand correctly your use case, I can think of 2 possible solutions:

1. Change your `file` field to accept multiple files.
   This is suitable if you don't need to store extra meta information with the different file "versions".
   For adding a new file versions - upload a new file to the field (see [Uploading files](https://pocketbase.io/docs/files-handling/#uploading-files); new uploaded files are appended to the file field array).
   For deleting old versions - remove a file entry from the field (see [Deleting files](https://pocketbase.io/docs/files-handling/#deleting-files)).
   
2. Create a separate `files` collection that will hold each canvas file.
   This is suitable if you want to provide extra meta information with the different file "versions".
   The collection could have the following sample structure:
   - `canvas` - single `relation` field pointing to the canvas record
   - `file`   - single `file` field
   - ... any other field needed for the extra meta information ...

   For adding new file versions - create a new record in the collection.
   For removing old file versions - remove a record from the collection.

For both approaches it is up to the developer to decide how the old versions will be managed and what their retention policy would be (_built-in auto delete for collection records eventually may get added in the future as it exists in my todo_).

**Comment by HEAVYPOLY** [comment]
with method 1, would loading become slower with the array of files if I only need to grab one of them at a time?

On S3 bucket it seems that every upload is saved.  Check this screenshot of a canvas record, all the files ever uploaded to it are stored even though it uses single file fields:
<img width="400" alt="image" src="https://github.com/pocketbase/pocketbase/assets/24432677/08df5a2a-b2d3-49b1-a65a-ce50374fbf02">

**Reply by ganigeorgiev** [reply]
I'll stop engage further in the discussion as I think we are going in circles. Please check https://github.com/pocketbase/pocketbase/discussions/3545#discussioncomment-7311976.

**Reply by HEAVYPOLY** [reply]
Ok, I will probably go with your proposed method #2 , it seems most flexible.  Thanks for answering.
